File: src/main/java/io/github/malczuuu/iemu/domain/FirmwareDTO.java
/*
 * Copyright (c) 2025 Damian Malczewski
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * SPDX-License-Identifier: MIT
 */
package io.github.malczuuu.iemu.domain;

import com.fasterxml.jackson.annotation.JsonIgnore;
import io.github.malczuuu.iemu.infrastructure.lwm2m.FirmwareUpdateDeliveryMethod;
import io.github.malczuuu.iemu.infrastructure.lwm2m.FirmwareUpdateResult;
import io.github.malczuuu.iemu.infrastructure.lwm2m.FirmwareUpdateState;

public class FirmwareDto {

  private final byte[] file;
  private final String fileChecksum;
  private final String packageUri;
  private final FirmwareUpdateState state;
  private final FirmwareUpdateResult result;
  private final String pkgVersion;
  private final FirmwareUpdateDeliveryMethod deliveryMethod;
  private final Integer progress;

  public FirmwareDto(
      byte[] file,
      String fileChecksum,
      String packageUri,
      FirmwareUpdateState state,
      FirmwareUpdateResult result,
      String pkgVersion,
      FirmwareUpdateDeliveryMethod deliveryMethod,
      Integer progress) {
    this.file = file;
    this.fileChecksum = fileChecksum;
    this.packageUri = packageUri;
    this.pkgVersion = pkgVersion;
    this.deliveryMethod = deliveryMethod;
    this.state = state;
    this.result = result;
    this.progress = progress;
  }

  @JsonIgnore
  public byte[] getFile() {
    return file;
  }

  public String getFileChecksum() {
    return fileChecksum;
  }

  public String getPackageUri() {
    return packageUri;
  }

  public FirmwareUpdateState getState() {
    return state;
  }

  public Integer getStateValue() {
    return state != null ? state.getValue() : null;
  }

  public FirmwareUpdateResult getResult() {
    return result;
  }

  public Integer getResultValue() {
    return result != null ? result.getValue() : null;
  }

  public String getPkgVersion() {
    return pkgVersion;
  }

  public FirmwareUpdateDeliveryMethod getDeliveryMethod() {
    return deliveryMethod;
  }

  public Integer getDeliveryMethodValue() {
    return deliveryMethod != null ? deliveryMethod.getValue() : null;
  }

  public Integer getProgress() {
    return progress;
  }
}

File: src/main/kotlin/io/github/malczuuu/iemu/domain/FirmwareDto.kt
package io.github.malczuuu.iemu.domain

import com.fasterxml.jackson.annotation.JsonIgnore
import io.github.malczuuu.iemu.infra.lwm2m.FirmwareUpdateDeliveryMethod
import io.github.malczuuu.iemu.infra.lwm2m.FirmwareUpdateResult
import io.github.malczuuu.iemu.infra.lwm2m.FirmwareUpdateState

data class FirmwareDto(
    @get:JsonIgnore val file: ByteArray?,
    val fileChecksum: String?,
    val packageUri: String?,
    val state: FirmwareUpdateState?,
    val result: FirmwareUpdateResult?,
    val pkgVersion: String?,
    val deliveryMethod: FirmwareUpdateDeliveryMethod?,
    val progress: Int?,
) {

  val stateValue: Int?
    get() = state?.value

  val resultValue: Int?
    get() = result?.value

  val deliveryMethodValue: Int?
    get() = deliveryMethod?.value

  // FIXME: remove file from DTO and drop equals and hashCode

  override fun equals(other: Any?): Boolean {
    if (this === other) return true
    if (javaClass != other?.javaClass) return false

    other as FirmwareDto

    if (progress != other.progress) return false
    if (!file.contentEquals(other.file)) return false
    if (fileChecksum != other.fileChecksum) return false
    if (packageUri != other.packageUri) return false
    if (state != other.state) return false
    if (result != other.result) return false
    if (pkgVersion != other.pkgVersion) return false
    if (deliveryMethod != other.deliveryMethod) return false
    if (stateValue != other.stateValue) return false
    if (resultValue != other.resultValue) return false
    if (deliveryMethodValue != other.deliveryMethodValue) return false

    return true
  }

  override fun hashCode(): Int {
    var result = progress ?: 0
    result = 31 * result + (file?.contentHashCode() ?: 0)
    result = 31 * result + (fileChecksum?.hashCode() ?: 0)
    result = 31 * result + (packageUri?.hashCode() ?: 0)
    result = 31 * result + (state?.hashCode() ?: 0)
    result = 31 * result + (this@FirmwareDto.result?.hashCode() ?: 0)
    result = 31 * result + (pkgVersion?.hashCode() ?: 0)
    result = 31 * result + (deliveryMethod?.hashCode() ?: 0)
    result = 31 * result + (stateValue ?: 0)
    result = 31 * result + (resultValue ?: 0)
    result = 31 * result + (deliveryMethodValue ?: 0)
    return result
  }
}
